fix: handle runtime instructions for multi-instance activities

chillleader opened the pull request:
## Description

This PR adds a check before executing the runtime instruction to determine if the element is running as part of a multi-instance activity.

If it is part of a multi-instance, we should not execute the runtime instruction for every individual instance of the activity. Instead, we should only execute it once when the whole multi-instance activity finishes (i.e. token leaves the element from the user's perspective).

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [x] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #39812

chillleader committed:
fix: handle runtime instructions for multi-instance

This is meant to fix a bug with runtime instructions:
without this check, the instruction logic doesn't work
correctly inside a multi-instance body.
Due to shared element IDs, the termination instruction was
executed for every invocation of a multi-instance element,
resulting in invalid state transitions/rejections in the log.
The correct behavior is to execute the instruction only once
when the whole multi-instance body completes.

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceWithTerminationInstructionTest.java:
```diff
@@ -271,4 +271,60 @@ record ->
+                        b.zeebeInputCollectionExpression("[1,2,3]")
+                            .zeebeInputElement("i")
+                            .zeebeOutputCollection("out")
+                            .zeebeOutputElementExpression("j + 2"))
```
🔧 This is not necessary for the test case. Let's remove it to focus the test case.

korthout reviewed (changes requested):
Good catch @chillleader 🚀 

👍 Generally looks like the right approach.
👍 I like the test case!

❌ I see some possible improvements to performance (not using a state lookup to determine whether this is an inner instance of a multi-instance body) and test accuracy (does the test case really check for the difference between multi-instance body and inner instance?)

Please have a look at my change requests and suggestions.

chillleader commented:
Thanks for reviewing @korthout, I applied your suggestions 👍

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceWithTerminationInstructionTest.java:
```diff
@@ -316,15 +312,19 @@ record ->
+            Tuple.tuple(
+                ProcessInstanceIntent.ELEMENT_COMPLETED, BpmnElementType.MULTI_INSTANCE_BODY),
+            Tuple.tuple(ProcessInstanceIntent.ELEMENT_TERMINATING, BpmnElementType.PROCESS),
+            Tuple.tuple(ProcessInstanceIntent.ELEMENT_TERMINATED, BpmnElementType.PROCESS));
```
💭 This change could've been a separate commit with it's own message explaining why.

korthout reviewed (approved):
Nice! 👏 

LGTM 👍 

🔧 I see opportunity to share more details about why these changes were needed in the commit messages. This helps future readers of the code to understand why it works the way it does. Please take a moment to expand your commit messages with more context and details before merging.

chillleader committed:
test: inline element id in test deployment

More concise syntax => easier to read. Functionality-wise,
the notations are equal.

chillleader committed:
test: clean up redundant preconditions in test

These conditions are not required in this test scenario and
can be removed safely.

chillleader committed:
test: add BPMN element type to assertion

This makes the test more precise, as we want to be sure that
the checked events are related to specific element types.

chillleader committed:
perf: avoid state lookup when checking runtime instructions

We need to do this check to verify we handle multi-instance
bodies correctly, but this can be done more efficiently as
already have the process model available in element. We can
therefore avoid an extra state lookup.
